jupyterhub/_config.py

```diff
@@ -0,0 +1,141 @@
+"""Utilites for getting the resolved Hub config
+
+Used in the /api/info endpoint
+"""
+
+from dataclasses import dataclass, field
+from typing import Optional
+
+from traitlets.log import get_logger
+
+from jupyterhub import crypto
+from jupyterhub.user import User
+
+
+@dataclass
+class _MockORMUser:
+    """Mock orm.User API, so we can instantiate a Spawner without a real user"""
+
+    id: int = 0
+    name: str = "username"
+    admin: bool = False
+    encrypted_auth_state: Optional[bytes] = None
+    state: dict = field(default_factory=dict)
+
+    # relationships likely to be used
+    orm_spawners: dict = field(default_factory=dict)
+    groups: list = field(default_factory=list)
+
+    # other fields technically exist, but shouldn't be accessed by Spawners
+    # these likely won't be exposed when we separate Spawners from db access
+    created = None
+    last_activity = None
+    cookie_id: str = ""
+    roles: list = field(default_factory=list)
+
+
+class _MockUser(User):
+    """Mock orm.User API, so we can instantiate a Spawner without a real user"""
+
+    def _new_orm_spawner(self, servername=""):
+        pass
+
+
+class _MockDB:
+    """Mock out deprecated db access"""
+
+    def add(self, obj):
+        pass
+
+    def commit(self):
+        pass
+
+
+def _mock_spawner(spawner_class, config):
+    user = _MockUser(orm_user=_MockORMUser(), db=_MockDB())
+    try:
+        return spawner_class(config=config)
+    except Exception:
+        return
+
+
+def get_resolved_config(hub, trait_filters=()):
+    """Return configuration as a dict
+
+    trait_filters is a sequence of strings of the form 'ClassName' or 'ClassName.trait_name'.
+    """
+    log = get_logger()
+    full_class_filters = set()
+    class_trait_filters = {}
+    for f in trait_filters:
+        if "." not in f:
+            full_class_filters.add(f)
+        else:
+            cls_name, _, trait_name = f.partition(".")
+            if cls_name not in full_class_filters:
+                cls_filters = class_trait_filters.setdefault(cls_name, set())
+                cls_filters.add(trait_name)
+
+    all_classes = full_class_filters | set(class_trait_filters.keys())
+
+    # track any sections we didn't look at via instances
+    unused_sections = set() | all_classes
```

Comment:
Isn't `all_classes` guaranteed to be a set ?
`full_class_filters` is one even if empty and if so, `all_classes` is a set of `class_trait_filters.keys`.